propagate_type_hints.py

```diff
@@ -0,0 +1,111 @@
+# Propagate type hints & signatures defined in _linear_operator.py to derived classes.
+# Here we leverage libcst which can preserve the original whitespace & formatting of the file
+# The idea is that we only want to change the type hints.
+# This way we can enforce consistency between the base class signature and derived signatures.
+
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import libcst as cst
+
+
+class TypingCollector(cst.CSTVisitor):
+    def __init__(self):
+        # stack for storing the canonical name of the current function
+        self.stack: List[Tuple[str, ...]] = []
+        # store the annotations
+        self.annotations: Dict[
+            Tuple[str, ...],  # key: tuple of canonical class/function name
+            Tuple[cst.Parameters, Optional[cst.Annotation]],  # value: (params, returns)
+        ] = {}
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> Optional[bool]:
+        self.stack.append(node.name.value)
+
+    def leave_ClassDef(self, node: cst.ClassDef) -> None:
+        self.stack.pop()
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> Optional[bool]:
+        self.stack.append(node.name.value)
+        self.annotations[tuple(self.stack)] = (node.params, node.returns)
+        return False  # pyi files don't support inner functions, return False to stop the traversal.
+
+    def leave_FunctionDef(self, node: cst.FunctionDef) -> None:
+        self.stack.pop()
+
+
+class TypingTransformer(cst.CSTTransformer):
+
+    # List of LinearOperator functions we do not want to propagate the signature from
+    excluded_functions = ["__init__", "_check_args", "__torch_function__"]
+
+    def __init__(self, annotations):
```

Comment:
Nit: maybe define an `Annotations` type that can be used for this arg

Reply:
OK. I have upgraded this to use an Annotations type as a TypedDict per [PEP 589](https://peps.python.org/pep-0589/)